### PR TITLE
GDC / MDC Overhaul

### PR DESCRIPTION
**Pull request by Page-Not-Found**
GDC / MDC Overhaul

The build is failing in my fork due to Xamarin account unavailability.

### CODE REVIEW

**Comment by 304NotModified** [comment]
Thanks @Page-Not-Found 

Please:

- rebase to lastest master to fix the build error - this is really easy with [Tortoisegit](https://tortoisegit.org/)

image



- Disable the warning CS0618: '....' is obsolete with pragma's (check the build log!)

**Comment by Page-Not-Found** [comment]
Thanks the branch is now successful.

**Comment by 304NotModified** [comment]
Well AppVeyor / Travis won't agree ;) 

Log4JXmlEventLayoutRenderer has been changed (wrongly).

> Disable the warning CS0618: '....' is obsolete with pragma's (check the build log!)

I see still all the warnings? 

```
LayoutRenderers\MDCTests.cs(52,13): warning CS0618: 'MappedDiagnosticsContext' is obsolete: 'Use NLog.Contexts.ThreadContext Instead.' [C:\projects\nlog\tests\NLog.UnitTests\NLog.UnitTests.sl4.csproj]
```

**Comment by 304NotModified** [comment]
:)

**Comment by 304NotModified** [comment]
Merged into branch! Good work ! :+1: :)

**Comment by UgurAldanmaz** [comment]
Why we remove `Get` method? I mean why `IContext` does not contains a `Get` method?

Every methods in `MappedDiagnosticsContext` calls the new `ThreadContext`, so we can do same thing for Get method. 

Yes the obsolete `Get` method returns value from `ThreadContext` by internally, but I think that we should support it at the `ThreadContext` class too. Should not we?

**Comment by 304NotModified** [comment]
We have the indexer? It doesn't add anything expect more complexity.  Do you agree?

**Comment by UgurAldanmaz** [comment]
Yes I agree. I just consider about backward compability.

File: src/NLog/Contexts/ThreadContext.cs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        }
+
+        #region [Enumerators]
+        /// <summary>
+        /// Returns an enumerator that iterates through Internal Dictionary structure
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+        {
+            var items = this.Dict;
```
Comment (UgurAldanmaz) [inline src/NLog/Contexts/ThreadContext.cs]: Do we need to clone `Dic` here? If yes, this line should be like this:

```
var items = new Dictionary<string, object>(this.Dict);
```
It works also ref types. My test is here: https://dotnetfiddle.net/ZNVqfL

What do you think?

Comment (304NotModified) [inline src/NLog/Contexts/ThreadContext.cs]: I don't think cloning in a `GetEnumerator` is common or advised isn't it? 



user call always clone if it they like.

**Pull request merged**